### PR TITLE
Allow overriding of the wrapper network timeout

### LINKED ISSUE #17575: Allow setting a custom timeout for gradle wrapper download
### Expected Behavior
It should be possible to set custom timeouts for the download of the wrapper jar

### Current Behavior
The connect and read timeouts are [hardcoded to 10s](https://github.com/gradle/gradle/blob/f54ff0afbd0b40a6d0cc4cc25d44e909fc1df0dc/subprojects/wrapper/src/main/java/org/gradle/wrapper/Download.java#L98..L99)

### Context
This causes us a problem sometimes, in our organisation we don't have direct access to maven central and so set a custom download URL in gradle-wrapper.properties to download from an internal artifact cache that proxies out to central.  When the artifact cache is busy (it has thousands of users, and runs anti virus scans etc) the read can timeout.

I'm happy to contribute a PR for this, however I'd like some advice on the preferred approach?

- Create new properties in gradle-wrapper.properties for the connection and read timeouts
- Read system properties for connection and read timeouts (like [other things](https://github.com/gradle/gradle/blob/f54ff0afbd0b40a6d0cc4cc25d44e909fc1df0dc/subprojects/wrapper/src/main/java/org/gradle/wrapper/Download.java#L188) do currently)

Any preference on naming for these properties?
Would it be acceptable to get this in both gradle v6 and v7 (again, happy to contribute)

Thanks!

### PR DESCRIPTION
**Pull request by goughy000**
Fixes #17575 

### Context
As discussed in the issue, experience in our own organisation shows that the default timeout of 10 seconds can result in build failures when the download source of the wrapper is responding slowly.  Rather than hardcode an arbitrary higher timeout, this PR allows users to set their own.

Based on the feedback in the issue triage, I have introduced a single `networkTimeout` property to gradle-wrapper.properties which is used for both Connection and Read timeouts.  This can also be set at wrapper init time like the other properties

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by goughy000** [comment]
@ljacomet any update on this..?

**Comment by ljacomet** [comment]
@goughy000 Sorry, this slipped through and did not get attention. Let me correct that.

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@goughy000 Any chance you can rebase this PR to a recent commit?
Some CI infra setup changed and this means running the build will fail with environment related failures.

**Comment by goughy000** [comment]
@ljacomet no problem, I will try find some time this weekend

File: subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java [hunk 1]
```diff
@@ -427,4 +432,30 @@ public PathBase getArchiveBase() {
...
+     *
+     * @since 7.3
+     */
+    @Nullable
+    @Optional
+    @Input
+    @Incubating
+    public String getNetworkTimeout() {
+        return networkTimeout == null ? null : String.valueOf(networkTimeout);
+    }
```
Comment (big-guy) [inline subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java:460]: We don't want to introduce more non-Provider API properties on tasks, so this needs to be a `Property<Integer>`. I think our `@Option` knows how to convert String -> Integer, but if it doesn't we can add that.

```suggestion
    @Internal
    @Incubating
    @Option(option = "network-timeout", description = "Timeout in ms to use when the wrapper is performing network operations")
    public Property<Integer> getNetworkTimeout() {
        return networkTimeout;
    }
```

`networkTimeout` would need to be initialized with:

```
private final Property<Integer> networkTimeout = getProject().getObjects().property(Integer.class);
```

In addition, in `WrapperPlugin`, we should set the conventional value for this near here: https://github.com/gradle/gradle/blob/2d0225d7406122d7e345f90607438e2ddd85cc07/subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/WrapperPlugin.java#L34

```
wrapper.getNetworkTimeout().convention(10000);
```

**Review by big-guy — CHANGES_REQUESTED** [review]
Thanks, this was actually really well done.

**Comment by ljacomet** [comment]
As Sterling commented, this PR is of good quality. Please take a look at his requests and assuming CI is happy with the result, this could be merged pretty soon.

**Comment by ljacomet** [comment]
@goughy000 Any chance you could pick this up to push it through the finish line?

**Pull request closed without merging**